include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_for.h

```diff
@@ -127,45 +151,48 @@ struct __parallel_for_large_submitter<__internal::__optional_kernel_name<_Name..
 
     // Once there is enough work to launch a group on each compute unit with our chosen __iters_per_item,
     // then we should start using this code path.
-    template <typename _ExecutionPolicy, typename _Fp>
+    template <typename _ExecutionPolicy, typename _Fp, typename... _Ranges>
     static std::size_t
     __estimate_best_start_size(const _ExecutionPolicy& __exec, _Fp __brick)
     {
+        using __params_t = __pfor_params<true /*__enable_tuning*/, _Fp, _Ranges...>;
         const std::size_t __work_group_size =
             oneapi::dpl::__internal::__max_work_group_size(__exec.queue(), __max_work_group_size);
         const std::uint32_t __max_cu = oneapi::dpl::__internal::__max_compute_units(__exec.queue());
-        return __work_group_size * _Fp::__preferred_iters_per_item * __max_cu;
+        return __work_group_size * __params_t::__iters_per_item * __max_cu;
     }
 
     template <typename _ExecutionPolicy, typename _Fp, typename _Index, typename... _Ranges>
     __future<sycl::event>
     operator()(_ExecutionPolicy&& __exec, _Fp __brick, _Index __count, _Ranges&&... __rngs) const
     {
+        using __params_t = __pfor_params<true /*__enable_tuning*/, _Fp, _Ranges...>;
         assert(oneapi::dpl::__ranges::__get_first_range_size(__rngs...) > 0);
         const std::size_t __work_group_size =
             oneapi::dpl::__internal::__max_work_group_size(__exec.queue(), __max_work_group_size);
         _PRINT_INFO_IN_DEBUG_MODE(__exec.queue());
         auto __event = __exec.queue().submit([__rngs..., __brick, __work_group_size, __count](sycl::handler& __cgh) {
             //get an access to data under SYCL buffer:
             oneapi::dpl::__ranges::__require_access(__cgh, __rngs...);
-            constexpr std::uint8_t __iters_per_work_item = _Fp::__preferred_iters_per_item;
-            constexpr std::uint8_t __vector_size = _Fp::__preferred_vector_size;
+            constexpr std::uint8_t __iters_per_work_item = __params_t::__iters_per_item;
+            constexpr std::uint8_t __vector_size = __params_t::__vector_size;
             const std::size_t __num_groups = oneapi::dpl::__internal::__dpl_ceiling_div(
                 __count, (__work_group_size * __vector_size * __iters_per_work_item));
             __cgh.parallel_for<_Name...>(
                 sycl::nd_range(sycl::range<1>(__num_groups * __work_group_size), sycl::range<1>(__work_group_size)),
                 [=](sycl::nd_item</*dim=*/1> __item) {
+                    __params_t __params;
                     const auto [__idx, __stride, __is_full] =
                         __stride_recommender(__item, __count, __iters_per_work_item, __vector_size, __work_group_size);
                     __strided_loop<__iters_per_work_item> __execute_loop{static_cast<std::size_t>(__count)};
                     if (__is_full)
```

Comment:
What about to add one more overload into `struct __strided_loop` for the case where nothing to do and calculate here proper param value in ternary operator + have only one call of `__execute_loop` without `if` calls?

Reply:
I added a similar suggestion to this. I think it is better to improve the safety of the `false_type{}` case of `__strided_loop` to avoid underflow for unsigned types so we do not need the outside `else if (__idx < __count)` check. With this change, we can use the ternary as suggested which is cleaner and future uses will be protected from the dangerous scenario where `__idx > __count`.

There is no change in performance.

Reply:
ok